eclair-core/src/main/scala/fr/acinq/eclair/io/ReconnectionTask.scala

```diff
@@ -0,0 +1,189 @@
+          case Peer.CONNECTED =>
+            log.info("peer is disconnected")
+            nextReconnectionDelay(initialDelay, nodeParams.maxReconnectInterval)
+        }
+        setTimer(RECONNECT_TIMER, TickReconnect, initialDelay, repeat = false)
+        goto(WAITING) using WaitingData(firstNextReconnectionDelay)
+      } else {
+        stay
+      }
+
+    case Event(FSM.Transition(_, Peer.DISCONNECTED, Peer.CONNECTED), _) =>
+      log.info("peer is connected")
+      stay
+  }
+
+  whenUnhandled {
+    case Event(_: Terminated, _) => stay
+
+    case Event(TickReconnect, _) => stay
+
+    case Event(FSM.Transition(_, Peer.INSTANTIATING, Peer.INSTANTIATING), _) => stay // instantiation transition
+
+    case Event(Peer.Connect(_, hostAndPort_opt), _) =>
+      // this happens completely independently of the automated reconnection process
+      // if we are already connecting/connected, the peer will kill any duplicate connections
+      hostAndPort_opt
+        .map(hostAndPort2InetSocketAddress)
+        .orElse(getPeerAddressFromDb(nodeParams.db.peers, nodeParams.db.network, remoteNodeId)) match {
+        case Some(address) => connect(address, Some(sender))
+        case None => sender ! "no address found"
+      }
+      stay
+  }
+
+  private def connect(address: InetSocketAddress, origin_opt: Option[ActorRef]): ActorRef = {
+    log.info(s"connecting to $address")
+    val connection = context.actorOf(Client.props(nodeParams, switchboard, router, address, remoteNodeId, origin_opt = origin_opt))
+    context.watch(connection)
+    connection
+  }
+
+  override def mdc(currentMessage: Any): MDC = {
+    Logs.mdc(Some(LogCategory.CONNECTION), Some(remoteNodeId))
+  }
+
+}
+
+object ReconnectionTask {
+
+  def props(nodeParams: NodeParams, remoteNodeId: PublicKey, switchboard: ActorRef, router: ActorRef): Props = Props(new ReconnectionTask(nodeParams, remoteNodeId, switchboard, router))
+
+  val RECONNECT_TIMER = "reconnect"
+
+  case object TickReconnect
+
+  // @formatter:off
+  sealed trait State
+  case object INIT extends State
+  case object CONNECTING extends State
+  case object WAITING extends State
+  case object IDLE extends State
+  // @formatter:on
+
+  // @formatter:off
+  sealed trait Data
+  case object Nothing extends Data
+  case class ConnectingData(connection: ActorRef, to: InetSocketAddress, nextReconnectionDelay: FiniteDuration) extends Data
+  case class WaitingData(nextReconnectionDelay: FiniteDuration) extends Data
+  // @formatter:on
+
+  def getPeerAddressFromDb(peersDb: PeersDb, networkDb: NetworkDb, remoteNodeId: PublicKey): Option[InetSocketAddress] = {
+    peersDb.getPeer(remoteNodeId) // TODO should we start with the network db which may be more up to date? or rotate?
+      .orElse(networkDb.getNode(remoteNodeId).flatMap(_.addresses.headOption)) // TODO gets the first of the list, improve selection?
```

Comment:
It would be interesting to grab the peer and network DBs from our node, and compare the addresses in both to see what discrepancies there are. It feels to me that if you're online most of the time, the peer DB should be more up-to-date than the network DB (provided that when our peer connects to us from a new address, we correctly save that address to the peer DB).

Doing a one-time analysis on those DBs will help us see what happens in the real world and should let us address these TODOs.

Reply:
Indeed, but if the peer changes ip the network db will be more accurate. There is some material for a future PR... 

> (provided that when our peer connects to us from a new address, we correctly save that address to the peer DB)

https://github.com/ACINQ/eclair/blob/split-peer/eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala#L107-L111

Reply:
> Indeed, but if the peer changes ip the network db will be more accurate

It depends...if our remote changes IP while it's connected, it will kill the TCP session but the remote should reconnect to us with the new IP very quickly. The `node_announcement` with its new IP will probably take longer to reach us. It's just a supposition though, looking at what's inside our DB with timestamps will probably help us learn how it behaves in the real world!

Reply:
> the remote should reconnect to us with the new IP very quickly

We only store the ip info for outgoing connections, because most of the time the peer isn't reachable.

Reply:
Oh yeah true, that makes sense